### PR TITLE
BFD-1456 Refactor - Introduce TransformerContext

### PR DESCRIPTION
**Pull request by cbrunefearless**
**JIRA Ticket:**
[BFD-1456](https://jira.cms.gov/browse/BFD-1456)

**User Story or Bug Summary:**

Tech debt code refactoring,  instead of passing just the metric registry, includetaxnumber boolean, and fdadrugdisplaycode, this refactored puts them in a transformer context object.
---

### What Does This PR Do?

Tech debt code refactoring,  instead of passing just the metric registry, includetaxnumber boolean, and fdadrugdisplaycode, this refactored puts them in a transformer context object.

### What Should Reviewers Watch For?

please check for good coding practices
If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X]  No
    * Does this PR modify or invalidate any of our security controls?
      * [ ]  Yes
      * [X]  No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ]  Yes
      * [X]  No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X]  No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

*[X]  I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X]  This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X]  All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X]  Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X]  Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/TransformerContext.java
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/TransformerContext.java:11]: these can all be final

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/TransformerContext.java
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/TransformerContext.java:21]: I think we can leave off the leading underscore with no ill effects.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/TransformerContext.java
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/TransformerContext.java:6]: missing the javadoc here and on the fields

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/TransformerContext.java
Comment (christrotteradhoc) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/TransformerContext.java:11]: I think private fields with getters is more consistent with the rest of the BFD codebase and would leave the door open for flexibility that we cannot anticipate.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/TransformerContext.java
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/TransformerContext.java:9]: I don't really think of the claim as being a part of the context as much as being the thing that is being transformed.

**Review by christrotteradhoc — COMMENTED** [review]
Thanks for getting this done. I have a few suggestions.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/TransformerContext.java
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/TransformerContext.java:40]: rather than referencing the short identifier, can we simply use `@return the {@link drugCodeDisplayLookup}` in these three getters?

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/DMEClaimTransformerV2.java
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/DMEClaimTransformerV2.java:52]: I think we ought to propagate the context down to this level too. I don't think it goes any further down, right?

Comment (christrotteradhoc) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/DMEClaimTransformerV2.java:52]: same comment elsewhere but let me know if you disagree

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/CarrierClaimTransformerV2.java
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/CarrierClaimTransformerV2.java:54]: could pass the context here

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/DMEClaimTransformerV2.java
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/DMEClaimTransformerV2.java:167]: could pass the context here

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/CarrierClaimTransformer.java
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/CarrierClaimTransformer.java:46]: could pass the context here

**Review by christrotteradhoc — APPROVED** [review]

**Review by mackec1 — APPROVED** [review]
Not going to hold this up, but I did provide a comment where it just seemed a little weird fetching a variable when the easiest wayt was to simply return true/false.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/DMEClaimTransformerV2.java
Comment (christrotteradhoc) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/DMEClaimTransformerV2.java:237]: I made a similar comment in one of the other threads and put it on the tech debt list as another item. I think we should not even store an optional in the context -- instead have the default value computed up front when we receive the config and keep the config in the context as a primitive and avoid this logic in the lower levels altogether.

**Comment by christrotteradhoc** [comment]
Chris - please make sure to adjust the PR title and commit message before merging to be a bit more standard and informative -- something like 'BFD-1456 Refactor by introducing TransformerContext'

**Pull request merged**
